flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/ExecNodeMetadataUtil.java

```diff
@@ -0,0 +1,294 @@
+        ExecNodeMetadata annotation = execNodeClass.getDeclaredAnnotation(ExecNodeMetadata.class);
+        if (annotation != null) {
+            metadata.add(annotation);
+        }
+
+        MultipleExecNodeMetadata annotations =
+                execNodeClass.getDeclaredAnnotation(MultipleExecNodeMetadata.class);
+        if (annotations != null) {
+            if (metadata.isEmpty()) {
+                for (ExecNodeMetadata annot : annotations.value()) {
+                    if (annot != null) {
+                        metadata.add(annot);
+                    }
+                }
+            } else {
+                throw new IllegalStateException(
+                        String.format(
+                                "ExecNode: %s is annotated both with %s and %s. Please use only "
+                                        + "%s or multiple %s",
+                                execNodeClass.getCanonicalName(),
+                                ExecNodeMetadata.class,
+                                MultipleExecNodeMetadata.class,
+                                MultipleExecNodeMetadata.class,
+                                ExecNodeMetadata.class));
+            }
+        }
+        return metadata;
+    }
+
+    private static void addToLookupMap(Class<? extends ExecNode<?>> execNodeClass) {
+        if (!JsonSerdeUtil.hasJsonCreatorAnnotation(execNodeClass)) {
+            throw new IllegalStateException(
+                    String.format(
+                            "ExecNode: %s does not implement @JsonCreator annotation on "
+                                    + "constructor.",
+                            execNodeClass.getCanonicalName()));
+        }
+
+        List<ExecNodeMetadata> metadata = extractMetadataFromAnnotation(execNodeClass);
+        if (metadata.isEmpty()) {
+            throw new IllegalStateException(
+                    String.format(
+                            "ExecNode: %s is missing %s annotation.",
+                            execNodeClass.getCanonicalName(),
+                            ExecNodeMetadata.class.getSimpleName()));
+        }
+
+        for (ExecNodeMetadata meta : metadata) {
+            doAddToMap(new ExecNodeNameVersion(meta.name(), meta.version()), execNodeClass);
+        }
+    }
+
+    private static void doAddToMap(
+            ExecNodeNameVersion key, Class<? extends ExecNode<?>> execNodeClass) {
+        if (LOOKUP_MAP.containsKey(key)) {
+            throw new IllegalStateException(
+                    String.format(
+                            "Found duplicate ExecNode: %s. This is a bug, please contact developers.",
+                            key));
+        }
+        LOOKUP_MAP.put(key, execNodeClass);
+    }
+
+    /**
+     * Returns the {@link ExecNodeMetadata} annotation of the class with the highest (most recent)
+     * {@link ExecNodeMetadata#version()}.
+     */
+    @SuppressWarnings("rawtypes")
```

Comment:
There are quite a few suppressions. As mentioned in a comment before, I think we can get rid of a couple of them by correct declaration.